---
title: "Module 2.1 - Logistic regressions"
author: "A. Gamble, I. Ali, N. Zaitlen, R. Wollman"
output: 
 html_document:
 toc: TRUE
 toc_float: TRUE
editor_options:
  chunk_output_type: console
---

## Setup

In this module, we will use the `ggplot2` package for plotting.

```{r setup, include = FALSE}

knitr::opts_chunk$set(echo = TRUE)

library(ggplot2) # plotting library

```

## Data exploration

### Loading and printing the data

In this module, as in Module 2.1, we will use the GeneExpression data we simulated in Module 1.2 Supplement, and visually explored in module 1.3. 

This data set contains information on gene expression (MeanCounts, representing measurements of protein quantities), gene transcription (MeanCountsTranscipt, representing measurements of mRNA quantities), and protein activity (MeanActivity, representing measurements of protein activity). Transcription levels are classically easier to measure (measured by RNA sequencing) than protein expression or activity levels (often measured by immuno-labelling such as immuno-histochemistry or flow-cytometry).

```{r dummydataimport}

# Load data
dummydata <- read.csv("Data/GeneExpressionData.csv", header = TRUE, sep = ",")

# Preview the data
head(dummydata)

```

In this module, we aim to measures the association between protein activity and the output of qualitative, binary test (returning 0 or 1, for non-detected or detected activity).

### Visualizing the data

```{r visualassessment}

# Use ggplot to create a scatter plot comparing transcriptions and expression levels
# length for the species
ggplot(data = dummydata, aes(x = MeanActivity, y = MeanActivityQuali)) +
  geom_point() +
  labs(x = "Protein Activity",
       y = "Activity Detection") 

# Add a logistic smooth line
ggplot(data = dummydata, aes(x = MeanActivity, y = MeanActivityQuali)) +
  geom_point() +
  geom_smooth(method = "glm", method.args = list(family = "binomial")) +
  labs(x = "Protein Activity",
       y = "Activity Detection") 

```

## Testing correlations with logistic regressions

Logistic regressions are used to measure the strength of association 
between a quantitative variable and a qualitative variable. Logistic regressions belong to generalized linear models. With such models the data are transformed before being fitted, and then use the same approach as classical logistic regressions. With logistical regressions (fit to binomial data), a logit function is applied to the response variable fit the model: `logit(Y) = log(Y / (1-Y))`. The inverse logit function can be used to back-transform predictions from the model: `intverse-logit(X) = 1 / (exp(-(a + b X)))`, with a and b the model estimates (parameters).
Generalized linear regressions are obtained with the `glm()` function.

```{r lm}

# glm() function to fit the generalized linear model to the data
# Here we use the 'binomial' family 
glm = glm(MeanActivityQuali ~ MeanActivity, data = dummydata, family = "binomial")

# Use summary to obtain the estimates for the parameters a and b, as well as the p-value
summary(glm)

# Extract the coefficient alpha and beta
coefficients(glm)

# Back-transform the coefficient with the inverse-logit function
exp(coefficients(glm))/(1+exp(coefficients(glm)))

```