---
title: "Module 2.1 - Correlations"
author: "A. Gamble, I. Ali, N. Zaitlen, R. Wollman"
output: 
 html_document:
 toc: TRUE
 toc_float: TRUE
editor_options:
  chunk_output_type: console
---

## Setup

In this module, we will use the `ggplot2` package for plotting.

```{r setup, include = FALSE}

knitr::opts_chunk$set(echo = TRUE)

library(ggplot2) # plotting library

```

## Data exploration

### Loading and printing the data

In this module, we will use the GeneExpression data we simulated in Module 1.2 Supplement, and visually explored in module 1.3. 

This data set contains information on gene expression (MeanCounts, representing measurements of protein quantities), gene transcription (MeanCountsTranscipt, representing measurements of mRNA quantities), and protein activity (MeanActivity, representing measurements of protein activity). Transcription levels are classically easier to measure (measured by RNA sequencing) than protein expression or activity levels (often measured by immuno-labelling such as immuno-histochemistry or flow-cytometry).

```{r dummydataimport}

# Using the read.csv() command import the GeneExpressionData.csv file
dummydata <- read.csv("Data/GeneExpressionData.csv", header = TRUE, sep = ",")

# View the data
View(dummydata)

```

In this module, we aim to assess whether those three variables are correlated or not. If those measures are correlated, then our experimenter could measure only transcription levels to predict expression and activity levels. 

Let's start by focusing on the association between gene expression and transcription levels.

### Visualizing the data

```{r visualassessment}

# Use ggplot to make a histogram to view the distribution of the transcriptions levels
ggplot(data = dummydata, aes(x = MeanCountsTranscript)) +
  geom_histogram() +
  labs(x = "Transcription level",
       y = "Counts") 

# Add a fill color to distinguish between genes
ggplot(data = dummydata, aes(x = MeanCountsTranscript, fill = GeneNames)) +
  geom_histogram() +
  labs(x = "Transcription level",
       y = "Counts") +
  scale_fill_discrete(name = "Gene")

# Use ggplot to create a scatter plot comparing transcriptions and expression levels
# length for the species
# Use scale_color_discrete to edit the name and labels for your legend
ggplot(data = dummydata, aes(x = MeanCountsTranscript, y = MeanCounts, color = GeneNames)) +
  geom_point() +
  labs(x = "Transcription level",
       y = "Expression level") +
  scale_color_discrete(name = "Gene")

```

## Testing correlations

### Pearson's correlation 

Pearson's correlation is used to measure the strength and direction of association 
between two quantitative variables. Pearson's coefficient corresponds to the ratio between the covariance of two variables and the product of their standard deviations. It is essentially a normalized measurement of the covariance, such that the result always has a value between âˆ’1 and 1. This measure can only reflect a **linear correlation** of variables.
Pearson's correlation coefficient can be obtained using the `cor()` function by setting to method to `pearson`. 

```{r pearson-cor}

# cor() function to calculate the correlation coefficient
cor(dummydata$MeanCountsTranscript, dummydata$MeanCounts, # Variables between which we want to measure the association
    method = c("pearson") # Method
    ) 

```

Correlation coefficients can also be obtained using the `cor.test()` function, which also returns a confidence interval of the correlation coefficient. 

```{r pearson-cor.test}

# cor.test() function to calculate the correlation coefficient
# and the associated confidence interval
cor.test(dummydata$MeanCountsTranscript, dummydata$MeanCounts, # Variables between which we want to measure the association
         method = "pearson" # Method
         )

```

### Spearman's rank correlation

Spearman's rank correlation is a non-parametric equivalent to Pearson's correlation. It is used to measure the strength and direction of association between two ranked quantitative variables. Spearman's correlation assesses **monotonic relationships (whether linear or not)**.
Spearman's rank correlation coefficient can also be obtained using the `cor()` function or the `cor.test()` function by setting to method to `spearman`.

```{r spearman}

# cor() function to calculate the correlation coefficient
cor(dummydata$MeanCountsTranscript, dummydata$MeanCounts, # Variables between which we want to measure the association
    method = c("spearman") # Method
    ) 

# cor.test() function to calculate the correlation coefficient
# and the associated confidence interval
cor.test(dummydata$MeanCountsTranscript, dummydata$MeanCounts, # Variables between which we want to measure the association
         method = "spearman" # Method
         )

```
Note that, as Spearman's method is based on ranks, it cannot compute exact p-value when there are tied values (i.e., when two are more observations are equal).

#### Exercise 1 

We now want to assess whether gene expression level is a good predictor of protein activity level. Repeat the same analysis to assess the association between MeanCounts (expression level) and MeanActivity (activity level). How do you interpret the results?

```{r exercise1}

# Histogram of MeanActivity, colored by GeneNames
# Make sure the axis names and legend are explicit



# Scatter plot of MeanActivity ~ MeanCounts, colored by GenesNames
# Make sure the axis names and legend are explicit



# Pearson's correlation



# Spearman's ranked correlation



# Notice the difference (or not) between the outcome of the two tests


# Interpretation
print('This analysis suggest that gene expression levels and protein activity levels are / are not correlated. Hence, gene expression level is / is not a good predictor of protein activity.')

```
## Testing correlation with permutations (extra acivity)

In this section, we are going to apply what we learned in Module 1.4 regarding permutations to the GeneExpression data set. As above, we aim to assess whether gene transcription, gene expression and protein activity levels are associated.

We start by writing a function that calculates the p-value and correlation coefficient using permutations. For additional explanations, go back to Module 1.4.

```{r permutations-function}

my_perm_corr_test = function(x, y, iter = 1000) {
  r_perm = array(dim = iter)
  for (i in 1:iter) { # seq(iter) same as 1:iter
    r_perm[i] = cor(x, sample(y))
  }
  r_real = cor(x, y)
  r_perm_abs = abs(r_perm)
  r_real_abs = abs(r_real)
  bigger_thans = r_real_abs > r_perm_abs
  p_grater_thans = mean(bigger_thans)
  pval = 1 - p_grater_thans
  rtn = list("r_perm" = r_perm,
             "pval" = pval,
             "r_real" = r_real)
  return(rtn)
}

```

We can now apply our `my_perm_corr_test()` to the data. 

```{r permutations}

# Execute the my_perm_corr_test() function
prmtst = my_perm_corr_test(dummydata$MeanCountsTranscript, dummydata$MeanCounts, iter = 1000)

# Visualize the distribution of the permutations
# The output of our permutation function is not a data frame so we must
# convert it to a data frame to plot in ggplot
ggplot(data.frame(prmtst), aes(r_perm)) +
  geom_histogram()

# Print the calculated p-value and correlation coefficient
prmtst$pval
prmtst$r_real 

```

#### Exercise 2

Repeat the same analysis to assess the association between MeanCounts (expression level) and MeanActivity (activity level). How do you interpret the results?

```{r exercise2}

# Execute the my_perm_corr_test() function



# Visualize the distribution of the permutations



# Print the calculated p-value and correlation coefficient



# Interpretation
print('This analysis suggest that gene expression levels and protein activity levels are / are not correlated. Hence, gene expression level is / is not a good predictor of protein activity.')

```

## Solutions to exercises

#### Exercise 1 

```{r exercise1solution}

# Histogram of MeanActivity, colored by GeneNames
# Make sure the axis names and legend are explicit
ggplot(data = dummydata, aes(x = MeanActivity, fill = GeneNames)) +
  geom_histogram() +
  labs(x = "Transcription level",
       y = "Counts") +
  scale_fill_discrete(name = "Gene")

# Scatter plot of MeanActivity ~ MeanCounts, colored by GenesNames
# Make sure the axis names and legend are explicit
ggplot(data = dummydata, aes(x = MeanCounts, y = MeanActivity, color = GeneNames)) +
  geom_point() +
  labs(x = "Transcription level",
       y = "Expression level") +
  scale_color_discrete(name = "Gene")

# Pearson's correlation
cor.test(dummydata$MeanCounts, dummydata$MeanActivity, 
         method = "pearson" 
         )

# Spearman's ranked correlation
cor.test(dummydata$MeanCounts, dummydata$MeanActivity, 
         method = "spearman" 
         )

# We prefer the function cor.test because it also report the confidence interval
# of the correlation coefficient

# Notice the difference (or not) between the outcome of the two tests

# Interpretation
print('This analysis suggest that gene expression levels and protein activity levels are correlated. Hence, gene expression level is a good predictor of protein activity.')

```

#### Exercise 2

```{r exercise2solution}

# Execute the my_perm_corr_test() function
prmtst = my_perm_corr_test(dummydata$MeanActivity, dummydata$MeanCounts, iter = 1000)

# Visualize the distribution of the permutations
ggplot(data.frame(prmtst), aes(r_perm)) +
  geom_histogram()

# Print the calculated p-value and correlation coefficient
prmtst$pval
prmtst$r_real 

# Interpretation
print('This analysis suggest that gene expression levels and protein activity levels are correlated. Hence, gene expression level is a good predictor of protein activity.')

```

## Going further

If you want more practice, you can use run similar analyses on the built-in Iris data set. The Iris data set was originally collected in 1890 by Sir Ronald Aylmer Fischer, also the person who invented the Fischer's exact test, maximum likelihood estimation and analysis of variance tests. Use of this data set is so common in statistical training, it has been developed into a built in package and data set that can be easily loaded and analyzed in R. This data set contains flower characteristics for three species of plants: setosa, virginica and versicolor. The characteristics include width and length characteristics of the flowers and sepals (the leaf under the flower). In total the data set contains 150 observations of 5 variables. You can for instance explore the correlation between petal length and petal width.