---
title: "Module 3.1 - Principal component analyses"
author: "A. Gamble, I. Ali, N. Zaitlen, R. Wollman"
output: 
 html_document:
 toc: TRUE
 toc_float: TRUE
editor_options:
  chunk_output_type: console
---

## Setup

In this module, we will use the `ggplot2` package for plotting.

```{r setup, include = FALSE}

knitr::opts_chunk$set(echo = TRUE)

library(tidyverse) 
library(ggplot2) # plotting library
library(factoextra)

```

## Data exploration

### Loading and printing the data

In this module, we will use to use the GeneExpression data we simulated in Module 1.2 Supplement, and visually explored in module 1.3. 

This data set contains information on gene expression (MeanCounts, representing measurements of protein quantities), gene transcription (MeanCountsTranscipt, representing measurements of mRNA quantities), and protein activity (MeanActivity, representing measurements of protein activity). Transcription levels are classically easier to measure (measured by RNA sequencing) than protein expression or activity levels (often measured by immuno-labelling such as immuno-histochemistry or flow-cytometry).


https://tavareshugo.github.io/data-carpentry-rnaseq

```{r dummydataimport}
trans_cts = read.csv("Data/RNAData/counts_transformed_subset.csv")
sample_info = read.csv("Data/RNAData/sample_info.csv")

trans_cts_long <- trans_cts %>% 
  pivot_longer(cols = wt_0_r1:mut_180_r3, 
               names_to = "sample", 
               values_to = "cts")

trans_cts_long <- full_join(trans_cts_long, sample_info, by = "sample")

ggplot(data = subset(trans_cts_long, gene == unique(trans_cts_long$gene)[1]), aes(x = minute, y = cts, color = replicate)) +
  geom_line() + 
  geom_point() +
  facet_grid(strain ~ . )

# Create a matrix from our table of counts
pca_matrix <- trans_cts %>% 
  # make the "gene" column become the rownames of the table
  column_to_rownames("gene") %>% 
  # coerce to a matrix
  as.matrix() %>% 
  # transpose the matrix so that rows = samples and columns = variables
  t()

pca_matrix[1:10, 1:5]

# Perform the PCA
pca <- prcomp(pca_matrix)

fviz_eig(pca)
```


```{r}
set.seed(101)
mat <- matrix(rnorm(nrow(decathlon2.active)*10), 10, 500)
## Warning in matrix(rnorm(nrow(decathlon2.active) * 10), 10, 500): data length
## [270] is not a sub-multiple or multiple of the number of columns [500]
# Perform PCA
myPCA <- prcomp(mat, scale = TRUE)

fviz_eig(myPCA)
```


```{r }
# you can use the fviz packages
fviz_pca_ind(pca,
             repel = TRUE     # Avoid text overlapping
             )

# or get the data and plot it yourself
pca_scale <- prcomp(pca_matrix, scale = TRUE)


# The PC scores are stored in the "x" value of the prcomp object
pc_scores <- as.data.frame(pca_scale$x)
pc_scores$sample = rownames(pc_scores)
pc_scores = pc_scores %>% relocate(sample)

# print the result
pc_scores

ggplot(data = pc_scores, aes(x = PC1, y = PC2)) +
  geom_point()

pc_scores_comp = full_join(pc_scores, sample_info, by = "sample")

ggplot(data = pc_scores_comp, aes(x = PC1, y = PC2, color = as.factor(minute), shape = strain)) +
  geom_point()

# Contributions of variables to PC1
fviz_contrib(pca_scale, choice = "var", axes = 1, top = 10)
# Contributions of variables to PC2
fviz_contrib(pca_scale, choice = "var", axes = 2, top = 10)

fviz_pca_var(pca_scale,
             col.var = "contrib", # Color by contributions to the PC
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping
             )


pc_loadings <- as.data.frame(pca_scale$rotation)
pc_loadings$gene = rownames(pc_loadings)
pc_loadings = pc_loadings %>% relocate(gene)

ggplot(data = pc_loadings) +
  geom_segment(aes(x = 0, y = 0, xend = PC1, yend = PC2), 
               arrow = arrow(length = unit(0.1, "in")),
               color = "brown") +
  geom_text(aes(x = PC1, y = PC2, label = gene),
            nudge_y = 0.005, size = 3) +
  scale_x_continuous(expand = c(0.02, 0.02))

fviz_pca_biplot(pca_scale, repel = TRUE,
                col.var = "#2E9FDF", # Variables color
                col.ind = "#696969"  # Individuals color
                )
```

### Exercise 1

Decathlon data

```{r}

data(decathlon2)
head(decathlon2)

decathlon2.active <- decathlon2[, 1:10]
res.pca <- prcomp(decathlon2.active, scale = TRUE)
fviz_eig(res.pca)

# you can use the fviz packages
fviz_pca_ind(res.pca,
             repel = TRUE     # Avoid text overlapping
             )

# or get the data and plot it yourself
res.pca <- prcomp(decathlon2.active, scale = TRUE)
new_PC_representation = res.pca$x
qplot(new_PC_representation[,1],new_PC_representation[,2],xlab = "PC-1",ylab="PC2")

# Contributions of variables to PC1
fviz_contrib(res.pca, choice = "var", axes = 1, top = 10)

# Contributions of variables to PC2
fviz_contrib(res.pca, choice = "var", axes = 2, top = 10)

fviz_pca_var(res.pca,
             col.var = "contrib", # Color by contributions to the PC
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping
             )


```


